luarules/gadgets/unit_water_depth_damage.lua

```diff
@@ -0,0 +1,173 @@
+function gadget:GetInfo()
+	return {
+		name = "Water Crush and Collision Damage",
+		desc = "Creates and handles water collision events, and kills units stuck underwater",
+		author = "SethDGamre",
+		date = "2024.9.22",
+		license = "GNU GPL, v2 or later",
+		layer = 0,
+		enabled = true
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+--use customParams.water_fall_damage_multiplier = 1.0 to change the amount of fall damage taken by specific units.
+
+--required velocity in a frame for a unit to take collision damage from falling into water.
+local velocityThreshold = 99 / Game.gameSpeed
+
+--any maxWaterDepth movedef equal to or above this number will not take drowning damage.
+-- performance optimisation to avoid checking amphs and hovers
+local isDrownableMaxWaterDepth = 5000
+
+--a percentage of health taken as damage per second when stuck below maxWaterDepth.
+local drowningDamage = 0.05
+
+--base damage percentage multiplied by velocity upon impact with water.
+local fallDamage = 0.18
+
+--this influences the compounding escalation of fall damage from water collisions.
+local fallDamageCompoundingFactor = 1.05
+
+--statics
+local gameFrame = 0
+local gameFrameExpirationThreshold = 3
+local gaiaTeamID = Spring.GetGaiaTeamID()
+local waterDamageDefID = Game.envDamageTypes.Water
+local gameSpeed = Game.gameSpeed
+
+--functions
+local spGetUnitIsDead = Spring.GetUnitIsDead
+local spValidUnitID = Spring.ValidUnitID
+local spAddUnitDamage = Spring.AddUnitDamage
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spGetUnitBasePosition = Spring.GetUnitBasePosition
+local spGetUnitPosition = Spring.GetUnitPosition
+local spSpawnCEG = Spring.SpawnCEG
+local spPlaySoundFile = Spring.PlaySoundFile
+local spTestMoveOrder = Spring.TestMoveOrder
+local spGetUnitHealth = Spring.GetUnitHealth
+local spDestroyUnit = Spring.DestroyUnit
+
+--tables
+local unitDefData = {}
+local transportDrops = {}
+local drowningUnitsWatch = {}
+local expiringTransportDrops = {}
+local livingTransports = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	local defData = {}
+	defData.fallDamageMultiplier = unitDef.customParams.water_fall_damage_multiplier or 1
+	defData.drowningDamage = unitDef.health * drowningDamage
+	defData.fallDamage = unitDef.health * fallDamage * defData.fallDamageMultiplier
+	defData.unitDefID = unitDefID
+	if unitDef.moveDef.depth then
+		if unitDef.moveDef.depth and unitDef.moveDef.depth >= isDrownableMaxWaterDepth then
+			if unitDef.moveDef.name and string.find(unitDef.moveDef.name, "hover") then --units must have "hover" in their movedef name in order to be treated as hovercraft
```

Comment:
I'm not familiar with the moveDef's, but this type of check overall seems risky to me. Why don't units just have these booleans defined with method calls instead of assigning them on the fly based on some arbitrary name?

Reply:
movedefs controls the spacing of a unit, where that unit can go against the terrain of the map. there are about 4 movedefs in the game for hover. there are far more hover unitdefs. all ground/naval units must have a movementclass defined (movedef).

Reply:
https://springrts.com/wiki/Movedefs.lua#MoveClass_Tags BAR units use the string method of putting "hover" somewhere in their name. In order to hover, it must have that string.

Unless there's a speedmod tag or whatever it's called I missed. I'll look into that. I was following Damgams advice

Reply:
use `unitDef.moveDef.smClass == Game.speedModClasses.Hover`

Reply:
that worked!